### DISCUSSION TITLE
Extraction of table data with usual and unsual structure in the pdf.

### DISCUSSION THREAD

**Discussion by Sylvester-Anthony** [jsvine/pdfplumber#1087] [Ask for help with specific PDFs]
Hello @jsvine , love the work with pdfplumber and I have been expirementing in extracting table data from pdfs , the problem is the pdfs have both both properly structured tables and tables that are not properly structured. I was able to extract all data from the table but the detection of columns was very irregular and my dataframe was something like this : 
<img width="1325" alt="Screenshot 2024-02-02 at 9 01 05 PM" src="https://github.com/jsvine/pdfplumber/assets/95081856/0a22d96a-f622-4453-b145-0620317333a0">
![Screenshot_3](https://github.com/jsvine/pdfplumber/assets/95081856/4be45b62-b90e-4a44-86be-48ce0bfacfac)

As you can see I am getting the data from the columns but the content after the first set of rows is not completely available. 

My end goal is to get all the data in full from the table, needn't exlucisvely be a data frame. I saw this piece of code from the other discussions where the full text gets detected: 

```
(
    im.reset()
    .draw_rects(p0.extract_words(keep_blank_chars=False))
    .draw_rects(p0.extract_words(keep_blank_chars=True), stroke="blue", fill=None)
)
```
And the result was all the text getting detected and this is exactly the end goal of mine : 

![Screenshot_2](https://github.com/jsvine/pdfplumber/assets/95081856/e15303cb-a0f1-410e-b83d-50329bae8e97)
![Screenshot_1](https://github.com/jsvine/pdfplumber/assets/95081856/1b7114ee-3147-48dc-8cd3-bb28639e53c1)

So my question is , is there a way to get all this detected text in any form because the end goal is to get all the data possible from a variety of pdfs not just one table. The expirement with the various page settings works for one table but it doesnt work for the other. 

The pdf is : 
[2022  Sustainability Report_NYSE_WM_2022.pdf](https://github.com/jsvine/pdfplumber/files/14142389/2022.Sustainability.Report_NYSE_WM_2022.pdf)

Looking forward towards your suggestions . Thank you !

-Sylvester

**Comment by jsvine** [comment]
Hi @Sylvester-Anthony, and thanks for the kind words about `pdfplumber`. Unfortunately, I'm not sure I fully understand your question. 

When you say, " is there a way to get all this detected text in any form," would something as simple as `page.extract_text(layout=True)` suffice? Or do you mean a more structured form?

**Comment by Sylvester-Anthony** [comment]
Thank you for the suggestion @jsvine ! I've got one more hiccup in here , if you have a look at page number 42 in the pdf , its a tabular structure and its similar to page number 43. The data in 43 gets detected while the data in page 42 is not considered a table while using page.extract_tables(). Any idea on what could be causing this and is there some settings I could look at ?

**Reply by jsvine** [reply]
My first suggestion would be to examine the results of `page.to_image().debug_tablefinder()` for both those pages. Do you see any major differences?

**Comment by Sylvester-Anthony** [comment]
The major differences I see is , in page 43 we can see the tables getting detected albeit not in whole but as parts but the table in page 42 is not getting detected at all. So my thought process was to extract content of pages where tables are detected but page 42 is not getting detected. 

![Image_1](https://github.com/jsvine/pdfplumber/assets/95081856/996e0d8a-74f0-4ae1-be8e-5f313a62d566)
![Image_2](https://github.com/jsvine/pdfplumber/assets/95081856/379c381c-a244-4d00-914c-913496e5e6bb)

**Reply by jsvine** [reply]
Thanks for flagging this @Sylvester-Anthony — unfortunately, it looks like you've run into a bug, which I've now opened an issue for: https://github.com/jsvine/pdfplumber/issues/1110

To get grab those horizontal lines in the meantime, you can try:

```python
horizontal_lines = (
    pdfplumber.table.merge_edges(
        pdfplumber.utils.filter_edges(page.edges, "h"),
        snap_x_tolerance=0,
        snap_y_tolerance=0,
        join_x_tolerance=-1,
        join_y_tolerance=0,
    )
)
```
